core/src/main/java/org/jruby/truffle/nodes/rubinius/TimePrimitiveNodes.java

```diff
@@ -0,0 +1,316 @@
+            return time.getNanoseconds();
+        }
+
+    }
+
+    @RubiniusPrimitive(name = "time_decompose")
+    public static abstract class TimeDecomposePrimitiveNode extends RubiniusPrimitiveNode {
+
+        @Child protected RubyTimeToDateTimeNode toDateTimeNode;
+
+        public TimeDecomposePrimitiveNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+            toDateTimeNode = new RubyTimeToDateTimeNode(context, sourceSection);
+        }
+
+        public TimeDecomposePrimitiveNode(TimeDecomposePrimitiveNode prev) {
+            super(prev);
+            toDateTimeNode = prev.toDateTimeNode;
+        }
+
+        @Specialization
+        public RubyArray timeDecompose(VirtualFrame frame, RubyTime time) {
+            final DateTime dateTime = toDateTimeNode.toDateTime(frame, time);
+            final Object[] decomposed = decompose(dateTime);
+            return new RubyArray(getContext().getCoreLibrary().getArrayClass(), decomposed, decomposed.length);
+        }
+
+        @CompilerDirectives.TruffleBoundary
+        public Object[] decompose(DateTime dateTime) {
+            final int sec = dateTime.getSecondOfMinute();
+            final int min = dateTime.getMinuteOfDay();
+            final int hour = dateTime.getHourOfDay();
+            final int day = dateTime.getDayOfMonth();
+            final int month = dateTime.getMonthOfYear();
+            final int year = dateTime.getYear();
+            final int wday = dateTime.getDayOfWeek();
+            final int yday = dateTime.getDayOfYear();
+            final Object isdst = getContext().getCoreLibrary().getNilObject();
+            final Object zone = getContext().getCoreLibrary().getNilObject();
+            return new Object[]{sec, min, hour, day, month, year, wday, yday, isdst, zone};
+        }
+
+    }
+
+    @RubiniusPrimitive(name = "time_strftime")
+    public static abstract class TimeStrftimePrimitiveNode extends RubiniusPrimitiveNode {
+
+        @Child protected RubyTimeToDateTimeNode toDateTimeNode;
+
+        public TimeStrftimePrimitiveNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+            toDateTimeNode = new RubyTimeToDateTimeNode(context, sourceSection);
+        }
+
+        public TimeStrftimePrimitiveNode(TimeStrftimePrimitiveNode prev) {
+            super(prev);
+            toDateTimeNode = prev.toDateTimeNode;
+        }
+
+        @Specialization
+        public RubyString timeStrftime(VirtualFrame frame, RubyTime time, RubyString format) {
+            return getContext().makeString(format(toDateTimeNode.toDateTime(frame, time), time.getNanoseconds(), format.getBytes()));
+        }
+
+
+        @CompilerDirectives.TruffleBoundary
+        public ByteList format(DateTime time, long nanoseconds, ByteList format) {
+            // TODO: converts everything to JRuby objects and back - should find a more direct way using ByteList
```

Comment:
This is entirely a personal thing, but I've found it very helpful to annotate TODOs with an author (initials or handle) and date.  Some will claim that git blame would be sufficient, but the moment someone else touches that line (e.g., moves it) both the author and the date look wrong unless you dig far enough back.  If we want to got this way, we could settle on a date format.  I've been annotating with a full date to avoid any ambiguity between US formats and whatever crazy thing you guys use.